src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,591 @@
+pub type Paths = [String; 1];
+
+#[derive(Debug)]
+pub struct FileWithFormat<F: FileSystem> {
+    pub handle: Arc<F::Handle>,
+    pub format: LogFileFormat,
+    pub path_id: PathId,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FileState {
+    pub first_seq: FileSeq,
+    pub total_len: usize,
+}
+
+pub struct FileList<F: FileSystem> {
+    /// Sequence number of the first file.
+    first_seq: FileSeq,
+    fds: VecDeque<FileWithFormat<F>>,
+}
+
+impl<F: FileSystem> FileList<F> {
+    #[inline]
+    pub fn new(first_seq: FileSeq, fds: VecDeque<FileWithFormat<F>>) -> Self {
+        let first_seq = if fds.is_empty() { 0 } else { first_seq };
+        Self { first_seq, fds }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.fds.len()
+    }
+
+    #[inline]
+    pub fn span(&self) -> (FileSeq, FileSeq) {
+        if !self.fds.is_empty() {
+            (self.first_seq, self.first_seq + self.fds.len() as u64 - 1)
+        } else {
+            (0, 0)
+        }
+    }
+
+    #[inline]
+    pub fn get(&self, file_seq: FileSeq) -> Result<Arc<F::Handle>> {
+        if !(self.first_seq..self.first_seq + self.fds.len() as u64).contains(&file_seq) {
+            return Err(Error::Corruption("file seqno out of range".to_owned()));
+        }
+        Ok(self.fds[(file_seq - self.first_seq) as usize]
+            .handle
+            .clone())
+    }
+
+    #[inline]
+    pub fn back(&self) -> Option<&FileWithFormat<F>> {
+        self.fds.back()
+    }
+
+    #[inline]
+    pub fn push_back(&mut self, file_seq: FileSeq, handle: FileWithFormat<F>) -> FileState {
+        if self.fds.is_empty() {
+            self.first_seq = file_seq;
+        }
+        self.fds.push_back(handle);
+        debug_assert_eq!(file_seq, self.first_seq + (self.fds.len() - 1) as FileSeq);
+        FileState {
+            first_seq: self.first_seq,
+            total_len: self.fds.len(),
+        }
+    }
+
+    #[inline]
+    pub fn pop_front(&mut self) -> Option<(FileSeq, FileWithFormat<F>)> {
+        match self.fds.pop_front() {
+            Some(fd) => {
+                let seq = self.first_seq;
+                self.first_seq += 1;
+                Some((seq, fd))
+            }
+            None => None,
+        }
+    }
+
+    /// Splits current file list.
+    ///
+    /// Returns the first splitted part of current file list.
+    #[inline]
+    pub fn split_by(&mut self, file_seq: FileSeq) -> (FileState, FileList<F>) {
+        let splitted_list =
+            if (self.first_seq..self.first_seq + self.fds.len() as u64).contains(&file_seq) {
+                let purged = file_seq.saturating_sub(self.first_seq) as usize;
+                let purged_file_list =
+                    FileList::new(self.first_seq, self.fds.drain(..purged).collect());
+                self.first_seq = file_seq;
+                purged_file_list
+            } else {
+                FileList::new(0, VecDeque::default())
+            };
```

Comment:
It's incorrect.
As the api `split_off` shows:
> Returns a newly allocated VecDeque. self contains elements [0, at), and the returned deque contains elements [at, len).
```Rust
use std::collections::VecDeque;

let mut buf: VecDeque<_> = [1, 2, 3].into();
let buf2 = buf.split_off(1);
assert_eq!(buf, [1]);
assert_eq!(buf2, [2, 3]);
```
This function will return an inverse result => as the example shows, what we expect should be:
> After splitted, buf => [2, 3], buf2 => [1].

Reply:
std::mem::swap(self, &mut buf2).

Reply:
Still not recommended. With this implementation, another corner case do not meet the requirement.
That is, when `file_seq` > `self.first_seq`, `split_by` should not return a splitted `FileList`, but `split_off` will return an unexpected `panic!`.
```Rust
// 11 12 13 14 15 => call split_by(17)
(1) current implementation => returns an empty FileList.
(2) this implementation => returns an (panic!) because of 17 - 11 > self.len()
```

Reply:
Then add a check.